pkg/kubelet/prober/worker.go

```diff
@@ -132,11 +132,15 @@ func (w *worker) run() {
 	probeTickerPeriod := time.Duration(w.spec.PeriodSeconds) * time.Second
 
 	// If kubelet restarted the probes could be started in rapid succession.
-	// Let the worker wait for a random portion of tickerPeriod before probing.
+	// Let the worker wait for tickerPeriod before probing.
 	// Do it only if the kubelet has started recently.
-	if probeTickerPeriod > time.Since(w.probeManager.start) {
-		time.Sleep(time.Duration(rand.Float64() * float64(probeTickerPeriod)))
+	noRunBefore := probeTickerPeriod
+	// Override: readiness worker should start not later than startup worker,
+	// otherwise it will miss the manual trigger of the probe once started.
+	if w.probeType == readiness && w.container.StartupProbe != nil {
+		noRunBefore = time.Duration(math.Min(float64(w.spec.PeriodSeconds), float64(w.container.StartupProbe.PeriodSeconds))) * time.Second
 	}
+	time.Sleep(noRunBefore - time.Since(w.probeManager.start))
```

Comment:
Could this be negative/null? What's the behaviour in that case?

Reply:
https://golang.org/src/time/sleep.go
> // A negative or zero duration causes Sleep to return immediately.